src/topics/async.md

```diff
@@ -0,0 +1,461 @@
+```
+
+Note how the stacktrace from the first call to `err()` does have the `run()`
+function in it, yet the second call doesn't. This is because the second call
+to `err()` happens _after_ `run()` has completed so `run()` is no longer part
+of the execution environment.
+
+This is where the event loop shows up: the second call to `err()` happens
+during a timeout which is handled in the event loop.
+
+
+#### Throwing and try/catch
+One thing to be careful about in asynchronous programming is throwing errors
+and attempting to catch them.
+
+Consider this example:
+
+```js
+function run() {
+    setTimeout(function() {
+        throw new Error('boom');
+    }, 100);
+}
+try {
+    run();
+}
+catch (err) {
+    // we'll never get here!
+    console.log('TROUBLE!', err.stack);
+}
+```
+
+Since the `run()` function returns quickly, the execution of the program moves
+outside of the `try` block. (In fact it moves to the _end_ of the file, but
+the program doesn't exit because there is a timeout waiting to happen.) Once
+the timeout happens the code is run which throws the error. Since the program
+has moved outside of the `try` block the `catch` doesn't catch this thrown
+error.
+
+It is because of this that exceptions are not commonly thrown from
+asynchronous functions. Instead a common practice is to pass the error object
+to the callback. (See "nodeback" above.)
+
+In desperation, you can register a global exception handler by listening for
+the `uncaughtException` event on the process.
+
+```js
+process.on('uncaughtException', function(err) {
+  console.error('TROUBLE!', err.stack);
+});
+```
+
+Since this event handler is called well outside of your algorithm there is
+often very little you can do to recover from this exception.
+
+Some Promise implementations will catch a thrown error and resolve the promise
+as failed.
+
+
+### Concurrency
+If using asynchronous programming to handle network requests, it is possible
+that more than one request is being processed at the same time. This means
+that global variables should be treated with care, and probably shouldn't be
+associated with a single request.
+
+Consider the following example:
+
+```js
+var server = require('http').createServer().listen(8000),
+    currentRequest;
+server.on('request', function(request, response) {
+    currentRequest = request;   // trouble set!
+    setTimeout(function() {
+        response.writeHead(200);
+        response.end('hello world');
+        currentRequest = null;  // trouble clear!
+    }, 1000);
+});
+```
+
+The way the `currentRequest` variable is used is problematic. When a second
+concurrent request comes in, the `trouble set!` line will clobber the first
+request object. Once the first request is done, the `trouble clear!` line will
+empty the variable even though there is still another request currently being
+processed.
+
+
+### Single Javascript Execution Thread
+The part of node.js which runs your javascript does so in a single thread.
```

Comment:
We should probably elaborate and explain that NodeJS isn't actually single threaded - and the difference between platform and user code.